service/history/replication/sync_state_retriever.go

```diff
@@ -0,0 +1,409 @@
+	switch err.(type) {
+	case nil:
+	case *serviceerror.NotFound:
+		s.logger.Info(fmt.Sprintf("SyncWorkflowState new run not found, newRunId: %v", newRunId),
+			tag.WorkflowNamespaceID(namespaceId.String()),
+			tag.WorkflowID(execution.WorkflowId),
+			tag.WorkflowRunID(execution.RunId))
+		return nil, nil
+	default:
+		return nil, err
+	}
+	versionHistory, err := versionhistory.GetCurrentVersionHistory(mu.GetExecutionInfo().VersionHistories)
+	if err != nil {
+		return nil, err
+	}
+	versionHistory = versionhistory.CopyVersionHistory(versionHistory)
+	releaseFunc(nil)
+	releaseFunc = nil
+	newRunEvents, err := s.getEventsBlob(ctx, versionHistory.BranchToken, common.FirstEventID, common.FirstEventID+1)
+	switch err.(type) {
+	case nil:
+	case *serviceerror.NotFound:
+		s.logger.Info(fmt.Sprintf("SyncWorkflowState new run event not found, newRunId: %v", newRunId),
+			tag.WorkflowNamespaceID(namespaceId.String()),
+			tag.WorkflowID(execution.WorkflowId),
+			tag.WorkflowRunID(execution.RunId))
+		return nil, nil
+	default:
+		return nil, err
+	}
+	if len(newRunEvents) == 0 {
+		s.logger.Info(fmt.Sprintf("SyncWorkflowState new run event is empty, newRunId: %v", newRunId),
+			tag.WorkflowNamespaceID(namespaceId.String()),
+			tag.WorkflowID(execution.WorkflowId),
+			tag.WorkflowRunID(execution.RunId))
+		return nil, nil
+	}
+	return &replicationpb.NewRunInfo{
+		RunId:      newRunId,
+		EventBatch: newRunEvents[0],
+	}, nil
+}
+
+func (s *SyncStateRetrieverImpl) getMutation(mutableState workflow.MutableState, versionedTransition *persistencepb.VersionedTransition) (*persistencepb.WorkflowMutableStateMutation, error) {
+	rootNode := mutableState.HSM()
+	updatedStateMachine, err := s.getUpdatedSubStateMachine(rootNode, versionedTransition)
+	if err != nil {
+		return nil, err
+	}
+	mutableStateClone := mutableState.CloneToProto()
+	err = workflow.SanitizeMutableState(mutableStateClone)
+	if err != nil {
+		return nil, err
+	}
+	if err := common.DiscardUnknownProto(mutableStateClone); err != nil {
+		return nil, err
+	}
+	tombstoneBatch := mutableStateClone.GetExecutionInfo().SubStateMachineTombstoneBatches
+	var tombstones []*persistencepb.StateMachineTombstoneBatch
+	for i, tombstone := range tombstoneBatch {
+		if workflow.CompareVersionedTransition(tombstone.VersionedTransition, versionedTransition) > 0 {
+			tombstones = tombstoneBatch[i:]
+			break
+		}
+	}
+	mutableStateClone.ExecutionInfo.UpdateInfos = nil
+	mutableStateClone.ExecutionInfo.SubStateMachinesByType = nil
+	mutableStateClone.ExecutionInfo.SubStateMachineTombstoneBatches = nil
+	var signalRequestedIds []string
+	if workflow.CompareVersionedTransition(mutableStateClone.ExecutionInfo.SignalRequestIdsLastUpdateVersionedTransition, versionedTransition) > 0 {
+		signalRequestedIds = mutableStateClone.SignalRequestedIds
+	}
+	return &persistencepb.WorkflowMutableStateMutation{
+		UpdatedActivityInfos:            getUpdatedInfo(mutableStateClone.ActivityInfos, versionedTransition),
```

Comment:
Should these fields be populated from ms.updateXXX or ms.pendingXXX?

Reply:
Fields in mutable state proto are copied from pending.

Reply:
I think ms.updateXXX is a subset of ms.pendingXXX. @yycptt could you confirm?

Reply:
yes. updateXXX is only used during a ms transaction and will be cleared when closing the transaction. 


> LastUpdateVersionedTransition is updated for ms.updateXXX instead of ms.pendingXXX

The pointers in those two maps point to the same thing.

Reply:
Thanks for explaination!